crates/rpc/rpc/src/eth/api/mod.rs

```diff
@@ -201,6 +207,74 @@ where
     }
 }
 
+impl<Provider, Pool, Network> EthApi<Provider, Pool, Network>
+where
+    Provider: BlockReaderIdExt + StateProviderFactory + EvmEnvProvider + 'static,
+    Pool: TransactionPool + Clone + 'static,
+    Network: Send + Sync + 'static,
+{
+    /// Configures the [CfgEnv] and [BlockEnv] for the pending block
+    ///
+    /// If no pending block is available, this will derive it from the `latest` block
+    pub(crate) fn pending_block_env_and_cfg(&self) -> EthResult<PendingBlockEnv> {
+        let origin = if let Some(pending) = self.provider().pending_block()? {
+            PendingBlockEnvOrigin::ActualPending(pending)
+        } else {
+            // no pending block from the CL yet, so we use the latest block and modify the env
+            // values that we can
+            let mut latest =
+                self.provider().latest_header()?.ok_or_else(|| EthApiError::UnknownBlockNumber)?;
+
+            // child block
+            latest.number += 1;
+            // assumed child block is in the next slot
+            latest.timestamp += 12;
+            // base fee of the child block
+            latest.base_fee_per_gas = latest.next_block_base_fee();
+
+            PendingBlockEnvOrigin::DerivedFromLatest(latest)
+        };
+
+        let mut cfg = CfgEnv::default();
+        let mut block_env = BlockEnv::default();
+        self.provider().fill_block_env_with_header(&mut block_env, origin.header())?;
+        self.provider().fill_cfg_env_with_header(&mut cfg, origin.header())?;
+
+        Ok(PendingBlockEnv { cfg, block_env, origin })
+    }
+
+    /// Returns the locally built pending block
+    pub(crate) async fn local_pending_block(&self) -> EthResult<Option<SealedBlock>> {
+        let pending = self.pending_block_env_and_cfg()?;
+        if pending.origin.is_actual_pending() {
+            return Ok(pending.origin.into_actual_pending())
+        }
+
+        // no pending block from the CL yet, so we need to build it ourselves via txpool
+        self.on_blocking_task(|this| async move {
```

Comment:
Makes me think we should probably add docs on which ops are blocking and when we recommend spawning blocking?

Reply:
good point, will do!